src/thread_safe_reference.cpp

```diff
@@ -0,0 +1,116 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "thread_safe_reference.hpp"
+
+#include "impl/realm_coordinator.hpp"
+
+#include <realm/util/scope_exit.hpp>
+
+using namespace realm;
+
+ThreadSafeReferenceBase::ThreadSafeReferenceBase(SharedRealm source_realm) : m_source_realm(source_realm)
+{
+    m_source_realm->verify_thread();
+    if (m_source_realm->is_in_transaction()) {
+        throw InvalidTransactionException("Cannot obtain thread safe reference during a write transaction.");
+    }
+
+    try {
+        m_version_id = get_source_shared_group().pin_version();
+    } catch (...) {
+        invalidate();
+        throw;
+    }
+}
+
+ThreadSafeReferenceBase::~ThreadSafeReferenceBase()
+{
+    if (!is_invalidated())
+        invalidate();
+}
+
+template <typename V, typename T>
+V ThreadSafeReferenceBase::invalidate_after_import(Realm& destination_realm, T construct_with_shared_group) {
+    REALM_ASSERT_DEBUG(destination_realm.thread_id() == std::this_thread::get_id());
```

Comment:
I think the idea here was that since this is an internal API, that we shouldn't verify the thread in release builds, and that we would catch any API misuse in debug builds/tests.